driver/driver.go

```diff
@@ -64,162 +77,114 @@ func NewRemote(ctx context.Context, j *job.Job, c cluster.Cluster, in input.Feed
+		for _, err := range s.Errors {
+			res.addErr(err)
 		}
-	}()
 
-	// 3.1. monitor status
-	// we use error channel for centralized lifecycle management
-	var (
-		activeStreamCnt = atomic.NewInt32(0)
-		errChan         = errchannel.New()
-		statusChan      = make(chan *lrmrpb.JobOutput)
-	)
-	defer errChan.Close()
+		asyncCtxCancel()
+	})
 
-	for host, stream := range streams {
-		host, stream := host, stream
-		activeStreamCnt.Add(1)
-		go func() {
+	// 2. async start jobs and connect streams
+	var wg sync.WaitGroup
+	for host, rpc := range m.Connections {
+		wg.Add(1)
+
+		go func(host string, rpc lrmrpb.NodeClient) {
 			defer func() {
 				if err := errorist.WrapPanic(recover()); err != nil {
-					errChan.Send(err)
-				}
-				cnt := activeStreamCnt.Dec()
-				if cnt == 0 {
-					close(statusChan)
+					res.addErr(err)
 				}
+
+				wg.Done()
 			}()
 
-			for ctx.Err() == nil {
+			req := &lrmrpb.StartJobRequest{
+				JobID: m.Job.ID,
+			}
+			stream, err := rpc.StartJobInForeground(asyncCtx, req)
+			if err != nil {
+				res.addErr(errors.Wrapf(err, "call StartJobInForeground to %s", host))
+				return
+			}
+
+			for asyncCtx.Err() == nil {
 				msg, err := stream.Recv()
 				if err != nil {
 					if err == io.EOF || status.Code(err) == codes.Canceled || errors.Is(err, context.Canceled) {
-						return
-					}
-					if status.Code(err) == codes.DeadlineExceeded {
-						errChan.Send(context.DeadlineExceeded)
-						return
+						// normal end of stream
+					} else if status.Code(err) == codes.DeadlineExceeded {
+						res.addErr(context.DeadlineExceeded)
+					} else {
+						res.addErr(errors.Wrapf(err, "receive status from %s", host))
 					}
-					errChan.Send(errors.Wrapf(err, "receive status from %s", host))
-					return
+					break
 				}
-				if activeStreamCnt.Load() > 0 {
-					statusChan <- msg
-				}
-			}
-		}()
-	}
-
-	// 3.2. handle status
-	jobMgr := job.NewLocalManager(m.Job)
-	jobMgr.OnJobCompletion(func(s *job.Status) {
-		if len(s.Errors) > 0 {
-			errChan.Send(s.Errors[0])
-			return
-		}
-		errChan.Send(nil)
-	})
 
-	var result []*lrdd.Row
-JobRun:
-	// 4. consume status stream
-	for {
-		select {
-		// 4.1. received status report message from executor node
-		case msg := <-statusChan:
-			if msg == nil {
-				// channel is closed before job completion
-				continue
-			}
-			switch msg.Type {
-			case lrmrpb.JobOutput_COLLECT_DATA:
-				result = append(result, msg.Data...)
-
-			case lrmrpb.JobOutput_REPORT_TASK_COMPLETION:
-				taskID := job.TaskID{
-					JobID:       m.Job.ID,
-					StageName:   msg.Stage,
-					PartitionID: msg.PartitionID,
-				}
-				if msg.TaskStatus == lrmrpb.JobOutput_FAILED {
-					_ = jobMgr.MarkTaskAsFailed(ctx, taskID, errors.New(msg.Error), msg.Metrics)
-					continue
+				// received status report message from executor node
+				if msg.Type == lrmrpb.JobOutput_COLLECT_DATA {
+					for _, row := range msg.Data {
+						res.rowChan <- row
+					}
```

Comment:
I think that the user cancels the job and doesn't call `result.Err` while this code pushes data to `res.rowChan`, this goroutine can be blocked forever. The below suggestion can prevent that situation.

```suggestion
				pushLoop:
					for _, row := range msg.Data {
						select {
						case <-asyncCtx.Done():
							break pushLoop
						case res.rowChan <- row:
						}
					}
```

Reply:
Done. https://github.com/ab180/lrmr/pull/14/commits/b48969f1b6b082d07e452545a8a11b63628107b7